sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Call.java

```diff
@@ -17,75 +17,295 @@
 
-  // TODO(damondouglas): remove suppress warnings when configuration utilized in future PR.
-  @SuppressWarnings({"unused"})
   private final Configuration<RequestT, ResponseT> configuration;
 
   private Call(Configuration<RequestT, ResponseT> configuration) {
     this.configuration = configuration;
   }
 
+  /**
+   * Sets the {@link SetupTeardown} to the {@link Call} {@link PTransform} instance. Checks for
+   * {@link SerializableUtils#ensureSerializable} serializable errors.
+   */
+  Call<RequestT, ResponseT> withSetupTeardown(SetupTeardown setupTeardown) {
+    setupTeardown = SerializableUtils.ensureSerializable(setupTeardown);
+    return new Call<>(configuration.toBuilder().setSetupTeardown(setupTeardown).build());
+  }
+
+  /**
+   * Overrides the default {@link #DEFAULT_TIMEOUT}. A {@link UserCodeTimeoutException} is thrown
+   * when {@link Caller#call}, {@link SetupTeardown#setup}, or {@link SetupTeardown#teardown} exceed
+   * the timeout.
+   */
+  Call<RequestT, ResponseT> withTimeout(Duration timeout) {
+    return new Call<>(configuration.toBuilder().setTimeout(timeout).build());
+  }
+
+  @Override
+  public Result<ResponseT> expand(PCollection<RequestT> input) {
+    TupleTag<ResponseT> responseTag = new TupleTag<ResponseT>() {};
+
+    Coder<RequestT> requestTCoder = checkStateNotNull(input.getCoder());
+    try {
+      requestTCoder.verifyDeterministic();
+    } catch (NonDeterministicException e) {
+      throw new RuntimeException(e);
+    }
+
+    PCollectionTuple pct =
+        input.apply(
+            CallFn.class.getSimpleName(),
+            ParDo.of(new CallFn<>(responseTag, configuration, requestTCoder))
+                .withOutputTags(responseTag, TupleTagList.of(FAILURE_TAG)));
+
+    return Result.of(configuration.getResponseCoder(), responseTag, pct);
+  }
+
+  private static class CallFn<RequestT, ResponseT> extends DoFn<RequestT, ResponseT> {
+    private final TupleTag<ResponseT> responseTag;
+    private final CallerWithTimeout<RequestT, ResponseT> caller;
+    private final SetupTeardownWithTimeout setupTeardown;
+
+    private transient @MonotonicNonNull ExecutorService executor;
+
+    private final Coder<RequestT> requestTCoder;
+
+    private CallFn(
+        TupleTag<ResponseT> responseTag,
+        Configuration<RequestT, ResponseT> configuration,
+        Coder<RequestT> requestTCoder) {
+      this.responseTag = responseTag;
+      this.requestTCoder = requestTCoder;
+      this.caller = new CallerWithTimeout<>(configuration.getTimeout(), configuration.getCaller());
+      this.setupTeardown =
+          new SetupTeardownWithTimeout(
+              configuration.getTimeout(), configuration.getSetupTeardown());
+    }
+
+    /**
+     * Invokes {@link SetupTeardown#setup} forwarding its {@link UserCodeExecutionException}, if
+     * thrown.
+     */
+    @Setup
+    public void setup() throws UserCodeExecutionException {
+      this.executor = Executors.newSingleThreadExecutor();
```

Comment:
I see what you mean. I was concerned that having more than one thread would cause surprises in the rate of elements processed. I am concerned that calling Web APIs that are particularly quota sensitive may be problematic if we execute as more than one thread. What do you think about this? What if we have an additional parameter that accepts an ExecutorService but default to the one assigned from `newFixedThreadPool(int nThreads, ThreadFactory threadFactory)` and we derive the ThreadFactory from defaultThreadFactory() and nThreads to 1?

Reply:
A larger thread pool will only help if we allow `DoFn`s to process more than one `request` at a given time. So we can just go with what you have now and may be extend in the future if we support invoking multiple parallel requests from the same `DoFn`.

Reply:
Adding https://github.com/apache/beam/issues/29248 for context.